opentelemetry-api/src/opentelemetry/correlationcontext/__init__.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2019, OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import itertools
+import string
+import typing
+from contextlib import contextmanager
+
+from opentelemetry.context import ContextAPI
+
+CORRELATION_CONTEXT_KEY = "correlation-context-key"
+
+
+class CorrelationContextManager:
+    """
+    Manages access to CorrelationContext entries
+    """
+
+    def __init__(
+        self, entries: typing.Optional[typing.Dict[str, typing.Any]] = None
+    ) -> None:
+        if entries:
+            self._correlation_context = copy.deepcopy(entries)
+        else:
+            self._correlation_context = {}
+
+    def get_correlations(self) -> typing.Dict:
+        """
+        Returns the entries in this CorrelationContext. The order of entries is not significant. Based
+        on the languagespecification, the returned value can be either an immutable collection or an
+        immutable iterator to the collection of entries in this CorrelationContext.
+
+        Returns:
+            An immutable iterator to the collection of entries in this CorrelationContext
+        """
+        return self._correlation_context.items()
+
+    def get_correlation(
+        self, context: ContextAPI, name: str
+    ) -> typing.Optional[typing.Any]:
+        """
+        To access the value for an entry by a prior event, the Correlations API provides a
+        function which takes a context and a name as input, and returns a value. Returns
+        the Value associated with the given Name, or null if the given Name is not present.
+
+        Args:
+            context: The context in which to find the CorrelationContext
+            name: The name of the entry to retrieve
+        Returns:
+            The value of the entry matching the name
+        """
+        # pylint: disable=no-self-use
+        correlation_context = context.get_value(CORRELATION_CONTEXT_KEY)
```

Comment:
Another surprising consequence of the snapshot/context disconnect is that `get_correlations` and `get_correlation` do such different things, e.g. that `ccm.get_correlations()[key]` and `ccm.get_correlation(key, context)` only get the same value if `context`'s values match the ones used to create `ccm`.